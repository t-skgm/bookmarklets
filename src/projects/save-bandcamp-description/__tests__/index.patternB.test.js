import { jest } from '@jest/globals';
import { JSDOM } from 'jsdom';
import { readFileSync } from 'fs';
import { resolve } from 'path';
// import { setTimeout } from 'timers/promises';
import { buildHtmlWithJsonLd, getDirname, mockedFetchFn } from '../../../../test/utils';

const readStubHtml = filename =>
  readFileSync(resolve(getDirname(import.meta.url), './__stub__/', filename), {
    encoding: 'utf8'
  });

describe('save-bandcamp-description', () => {
  describe('Album page B (without tags, location)', () => {
    const albumHtmlB = readStubHtml('bandcamp_album_b.html');

    /** @type {JSDOM} */
    let dom;
    /** @type {Window} */
    let window;
    /** @type {Document} */
    let document;

    beforeAll(() => {
      // init window/document with jsdom
      dom = new JSDOM(albumHtmlB);
      global.document = document = dom.window.document;
      global.window = window = dom.window;
      global.DOMParser = dom.window.DOMParser;

      // mock fetch
      global.fetch = jest.fn(mockedFetchFn(buildHtmlWithJsonLd));
    });

    test('test jsdom mocking', async () => {
      const documentTitle = document.querySelector('title').textContent;
      expect(documentTitle).toMatchInlineSnapshot(`"The Warehouse Demos | Blewett"`);
    });

    describe('execute script', () => {
      let linkClickSpy;
      let createObjectURLSpy;

      beforeAll(async () => {
        // set spy fn
        linkClickSpy = jest.spyOn(window.HTMLElement.prototype, 'click');
        createObjectURLSpy = jest.spyOn(global.URL, 'createObjectURL');

        // execute script
        await import('../index');
      });

      test('alert never to be called', () => {
        expect(global.alert).not.toBeCalled();
      });

      test('download file (click link) has fired 1 times', () => {
        expect(linkClickSpy).toBeCalledTimes(1);
      });

      test('generated album info text', () => {
        const textData = createObjectURLSpy.mock.calls[0][0].blobParts[0];

        expect(textData).toMatchSnapshot();
      });
    });
  });
});
