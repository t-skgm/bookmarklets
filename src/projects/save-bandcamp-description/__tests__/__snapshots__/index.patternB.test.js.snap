// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`save-bandcamp-description Album page B (without tags, location) execute script generated album info text 1`] = `
"Blewett
The Warehouse Demos
https://blewett.bandcamp.com/album/the-warehouse-demos

<Track list>
1. So You Say
2. Child Asthma
3. Across State Senate
4. On The Back Burner
5. Allegro
6. Right Handed

<Description>
Recorded in an abandoned coffin factory in Fond du Lac, WI between 11/2002 & 2/2003.

<Credits>
Matt Duel - Guitar/Vocals
Justin Kibbel - Bass
Kevin Duel - Drums

Sean Wilke - screams on "On The Back Burner"

<Published>
2003-02-02

<Lisence>
Attribution Non Commercial No Derivatives

<Tags>
-

<Publisher>
[name]
Blewett"
`;
